File: backend/app/schemas/comparison.py
from typing import List, Optional
from pydantic import BaseModel, ConfigDict, Field


class ComparisonRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)
    drug1_id: str = Field(..., validation_alias="drug1Id", serialization_alias="drug1Id")
    drug2_id: str = Field(..., validation_alias="drug2Id", serialization_alias="drug2Id")


class ComparisonCitation(BaseModel):
    citation_id: str = Field(..., serialization_alias="citationId")
    document_id: str = Field(..., serialization_alias="documentId")
    document_name: str = Field(..., serialization_alias="documentName")
    page: int = 0
    section: Optional[str] = None
    text: Optional[str] = None
    score: Optional[float] = None


class ComparisonCell(BaseModel):
    content: str
    citations: List[ComparisonCitation] = []
    status: str = "normal"


class ComparisonAttribute(BaseModel):
    key: str
    label: str
    drug1: ComparisonCell
    drug2: ComparisonCell


class DrugInfo(BaseModel):
    id: str
    name: str
    generic_name: Optional[str] = Field(default=None, serialization_alias="genericName")
    drug_class: Optional[str] = Field(default=None, serialization_alias="drugClass")
    document_name: str = Field(..., serialization_alias="documentName")
    page_count: int = Field(..., serialization_alias="pageCount")


class ComparisonSummary(BaseModel):
    total_attributes: int = Field(..., serialization_alias="totalAttributes")
    warning_count: int = Field(..., serialization_alias="warningCount")
    highlight_count: int = Field(..., serialization_alias="highlightCount")
    unavailable_count: int = Field(..., serialization_alias="unavailableCount")
    both_unavailable_count: int = Field(..., serialization_alias="bothUnavailableCount")


class ComparisonResult(BaseModel):
    drug1: DrugInfo
    drug2: DrugInfo
    attributes: List[ComparisonAttribute]
    summary: ComparisonSummary

File: frontend/src/types/comparison.ts
import type { Citation } from './chat'

export type ComparisonCellStatus = 'normal' | 'warning' | 'highlight' | 'unavailable'

export type ComparisonCitation = Citation

export interface ComparisonCell {
  content: string
  citations: ComparisonCitation[]
  status?: ComparisonCellStatus
}

export interface ComparisonAttribute {
  key: string
  label: string
  drug1: ComparisonCell
  drug2: ComparisonCell
}

export interface DrugInfo {
  id: string
  name: string
  genericName?: string
  drugClass?: string
  documentName?: string
  pageCount?: number
}

export interface ComparisonSummary {
  totalAttributes: number
  warningCount: number
  highlightCount: number
  unavailableCount: number
  bothUnavailableCount: number
}

export interface ComparisonResult {
  drug1: DrugInfo
  drug2: DrugInfo
  attributes: ComparisonAttribute[]
  summary?: ComparisonSummary
}

export interface SavedComparison {
  id: string
  title: string
  drug1Id: string
  drug2Id: string
  drug1Name: string
  drug2Name: string
  savedAt: string
  notes?: string
  result: ComparisonResult
}

export const COMPARISON_ATTRIBUTE_KEYS = [
  'indications',
  'dosage_administration',
  'warnings',
  'contraindications',
  'drug_interactions',
  'adverse_reactions',
  'use_in_specific_populations',
  'pregnancy',
  'pediatric_use',
  'geriatric_use',
  'renal_impairment',
  'hepatic_impairment',
  'storage',
] as const

export type ComparisonAttributeKey = (typeof COMPARISON_ATTRIBUTE_KEYS)[number]

export const COMPARISON_ATTRIBUTE_LABELS: Record<ComparisonAttributeKey, string> = {
  indications: 'Indications',
  dosage_administration: 'Dosage & Administration',
  warnings: 'Warnings',
  contraindications: 'Contraindications',
  drug_interactions: 'Drug Interactions',
  adverse_reactions: 'Adverse Reactions',
  use_in_specific_populations: 'Use in Specific Populations',
  pregnancy: 'Pregnancy',
  pediatric_use: 'Pediatric Use',
  geriatric_use: 'Geriatric Use',
  renal_impairment: 'Renal Impairment',
  hepatic_impairment: 'Hepatic Impairment',
  storage: 'Storage',
}
